File: OctoberChallenge/week1/combination-sum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        if(candidates.size() == 0) return {{}};

        // Sort required: O(nlgn)
        sort(candidates.begin(), candidates.end());

        vector<int> tmp;
        vector<vector<int>> res;
        findCandidateRecursive(candidates, target, res, tmp, 0);

        return res;
    }

    void findCandidateRecursive(vector<int>& candidates, int target,
                                vector<vector<int>>& res, vector<int>& tmp, int i) {

        if(target < 0) return; //early return

        if(target == 0) { //earlier return after update res once
            res.push_back(tmp);
            return;
        }

        // General case:
        while(i < candidates.size() && target-candidates[i] >= 0) {
            tmp.push_back(candidates[i]);

            findCandidateRecursive(candidates, target-candidates[i], res, tmp, i);

            i++;

            tmp.pop_back();
        }
    }
};

File: OctoberChallenge/week1/combination-sum.py
class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        if len(candidates) == 0:
            return [[]];

        ## Member variables & methods in current scope
        res, tmp = [[]], [];
        def findCandidatesRecursive(lst, s, i):
            """ Recursive & back-tracking """
            if s < 0:
                return None;

            if s == 0:
                res.append(tmp);
                return None;

            while i < len(lst) and s-lst[i] >= 0:
                tmp.append(lst[i]);
                findCandidatesRecursive(lst, s-lst[i], i);
                i += 1;
                tmp.pop();

        candidates.sort(); #O(nlgn)
        findCandidatesRecursive(candidates, target, 0); #side effects only
        return res;
